Tests failing with pip 24.0

```console
$ python --version
Python 3.11.9
$ pip --version
pip 24.0 from /home/sudhackar/.pyenv/versions/3.11.9/envs/venv-soss/lib/python3.11/site-packages/pip (python 3.11)
$ tox p -e py311-linux
.pkg: recreate env because python changed version_info=[3, 12, 3, 'final', 0]->[3, 11, 9, 'final', 0] | executable='/usr/bin/python3.12'->'/home/sudhackar/.pyenv/versions/3.11.9/bin/python3.11'
...
======================================================================
FAIL: test_pip_install (__main__.SetuptoolsWheel.test_pip_install)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/python-versioneer-upstream/test/git/test_invocations.py", line 465, in test_pip_install
    self.run_in_venv(venv, venv,
  File "/dev/shm/python-versioneer-upstream/test/git/test_invocations.py", line 85, in run_in_venv
    return self.command(*bin_args, *args, workdir=workdir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/dev/shm/python-versioneer-upstream/test/git/common.py", line 16, in command
    self.fail("problem running command %s" % cmd)
AssertionError: problem running command /tmp/tmp5aim3n9h/venvs/setuptools-wheel-pip-install/bin/python

----------------------------------------------------------------------
Ran 28 tests in 250.710s

FAILED (failures=15)
  py311-linux: FAIL code 1 (316.33=setup[4.39]+cmd[0.15,0.07,0.33,0.05,60.55,250.80] seconds)
  evaluation failed :( (316.36 seconds)
```

pinning pip version to something like 23.2.1 will make the tests pass

```diff
diff --git a/test/git/test_invocations.py b/test/git/test_invocations.py
index 1f61ffb..e7e6b5a 100644
--- a/test/git/test_invocations.py
+++ b/test/git/test_invocations.py
@@ -58,6 +58,8 @@ class _Invocations(common.Common):
             self.run_in_venv(venv_dir, venv_dir,
                              'pip', 'install', '-U',
                              'pip', 'wheel', 'packaging')
+            self.run_in_venv(venv_dir, venv_dir, 'pip', 'install',
+                             'pip==23.2.1', '--upgrade')
         if sys.version_info.major == 3 and sys.version_info.minor >= 12:
             # Py 3.12 no longer default-installs setuptools in environments
             # See PEP-632
```

will pass

```
$ tox p -e py311-linux
  py311-linux: OK (492.07=setup[2.03]+cmd[0.14,0.06,0.13,0.04,60.69,428.31,0.13,0.06,0.05,0.07,0.24,0.11] seconds)
  congratulations :) (492.10 seconds)
```

--- a/test/git/test_invocations.py
+++ b/test/git/test_invocations.py
@@ -109,7 +109,7 @@ class _Invocations(common.Common):
     def make_demolib_sdist(self):
         # create an sdist of demolib-1.0 . for the *lib*, we only use the
         # tarball, never the repo.
-        demolib_sdist = self.subpath("cache", "demolib-1.0.tar")
+        demolib_sdist = self.subpath("cache", "demolib-1.0.tar.gz")
         if os.path.exists(demolib_sdist):
             return demolib_sdist
         libdir = self.subpath("build-demolib")
@@ -120,8 +120,8 @@ class _Invocations(common.Common):
         self.git("add", "--all", workdir=libdir)
         self.git("commit", "-m", "comment", workdir=libdir)
         self.git("tag", "demolib-1.0", workdir=libdir)
-        self.python("setup.py", "sdist", "--format=tar", workdir=libdir)
-        created = os.path.join(libdir, "dist", "demolib-1.0.tar")
+        self.python("setup.py", "sdist", "--format=gztar", workdir=libdir)
+        created = os.path.join(libdir, "dist", "demolib-1.0.tar.gz")
         self.assertTrue(os.path.exists(created))
         shutil.copyfile(created, demolib_sdist)
         return demolib_sdist
@@ -163,7 +163,7 @@ class _Invocations(common.Common):
         if os.path.exists(unpack_into):
             shutil.rmtree(unpack_into)
         os.mkdir(unpack_into)
-        with tarfile.TarFile(sdist) as t:
+        with tarfile.open(sdist, "r:gz") as t:
             t.extractall(path=unpack_into)
         unpacked = os.path.join(unpack_into, "demoappext-2.0")
         self.assertTrue(os.path.exists(unpacked))
@@ -172,12 +172,12 @@ class _Invocations(common.Common):
     def make_setuptools_extension_sdist(self):
         # create an sdist tarball of demoappext-setuptools at 2.0
         demoappext_setuptools_sdist = self.subpath("cache", "setuptools",
-                                                   "demoappext-2.0.tar")
+                                                   "demoappext-2.0.tar.gz")
         if os.path.exists(demoappext_setuptools_sdist):
             return demoappext_setuptools_sdist
         repodir = self.make_setuptools_extension_repo()
-        self.python("setup.py", "sdist", "--format=tar", workdir=repodir)
-        created = os.path.join(repodir, "dist", "demoappext-2.0.tar")
+        self.python("setup.py", "sdist", "--format=gztar", workdir=repodir)
+        created = os.path.join(repodir, "dist", "demoappext-2.0.tar.gz")
         self.assertTrue(os.path.exists(created), created)
         shutil.copyfile(created, demoappext_setuptools_sdist)
         return demoappext_setuptools_sdist
@@ -215,24 +215,24 @@ class _Invocations(common.Common):
     def make_setuptools_sdist(self):
         # create an sdist tarball of demoapp2-setuptools at 2.0
         demoapp2_setuptools_sdist = self.subpath("cache", "setuptools",
-                                                 "demoapp2-2.0.tar")
+                                                 "demoapp2-2.0.tar.gz")
         if os.path.exists(demoapp2_setuptools_sdist):
             return demoapp2_setuptools_sdist
         repodir = self.make_setuptools_repo()
-        self.python("setup.py", "sdist", "--format=tar", workdir=repodir)
-        created = os.path.join(repodir, "dist", "demoapp2-2.0.tar")
+        self.python("setup.py", "sdist", "--format=gztar", workdir=repodir)
+        created = os.path.join(repodir, "dist", "demoapp2-2.0.tar.gz")
         self.assertTrue(os.path.exists(created), created)
         shutil.copyfile(created, demoapp2_setuptools_sdist)
         return demoapp2_setuptools_sdist
 
     def make_setuptools_sdist_subproject(self):
         demoapp2_setuptools_sdist = self.subpath("cache", "setuptools",
-                                                 "demoapp2-subproject-2.0.tar")
+                                                 "demoapp2-subproject-2.0.tar.gz")
         if os.path.exists(demoapp2_setuptools_sdist):
             return demoapp2_setuptools_sdist
         projectdir = self.make_setuptools_repo_subproject()
-        self.python("setup.py", "sdist", "--format=tar", workdir=projectdir)
-        created = os.path.join(projectdir, "dist", "demoapp2-2.0.tar")
+        self.python("setup.py", "sdist", "--format=gztar", workdir=projectdir)
+        created = os.path.join(projectdir, "dist", "demoapp2-2.0.tar.gz")
         self.assertTrue(os.path.exists(created), created)
         shutil.copyfile(created, demoapp2_setuptools_sdist)
         return demoapp2_setuptools_sdist
@@ -243,7 +243,7 @@ class _Invocations(common.Common):
         if os.path.exists(unpack_into):
             shutil.rmtree(unpack_into)
         os.mkdir(unpack_into)
-        with tarfile.TarFile(sdist) as t:
+        with tarfile.open(sdist, "r:gz") as t:
             t.extractall(path=unpack_into)
         unpacked = os.path.join(unpack_into, "demoapp2-2.0")
         self.assertTrue(os.path.exists(unpacked))
@@ -255,7 +255,7 @@ class _Invocations(common.Common):
         if os.path.exists(unpack_into):
             shutil.rmtree(unpack_into)
         os.mkdir(unpack_into)
-        with tarfile.TarFile(sdist) as t:
+        with tarfile.open(sdist, "r:gz") as t:
             t.extractall(path=unpack_into)
         unpacked = os.path.join(unpack_into, "demoapp2-2.0")
         self.assertTrue(os.path.exists(unpacked))
@@ -379,13 +379,13 @@ class SetuptoolsRepo(_Invocations, unittest.TestCase):
     def test_sdist(self):
         sdist = self.make_setuptools_sdist() # asserts version as a side-effect
         # make sure we used setuptools/sdist, not distutils/sdist
-        with tarfile.TarFile(sdist) as t:
+        with tarfile.open(sdist, "r:gz") as t:
             self.assertIn("demoapp2-2.0/src/demoapp2.egg-info/PKG-INFO", t.getnames())
 
     def test_sdist_subproject(self):
         sdist = self.make_setuptools_sdist_subproject()
         # make sure we used setuptools/sdist, not distutils/sdist
-        with tarfile.TarFile(sdist) as t:
+        with tarfile.open(sdist, "r:gz") as t:
             self.assertIn("demoapp2-2.0/src/demoapp2.egg-info/PKG-INFO", t.getnames())
 
     def test_pip_install(self):

